tests/node_remediation_test.go

```diff
@@ -21,24 +21,111 @@ package tests_test
 
 import (
 	"flag"
+	"time"
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	v1 "kubevirt.io/node-recovery/pkg/apis/noderecovery/v1alpha1"
 	"kubevirt.io/node-recovery/pkg/client"
+	"kubevirt.io/node-recovery/pkg/controller"
 	"kubevirt.io/node-recovery/tests"
 )
 
 var _ = Describe("Node Remediation", func() {
+	var sshExecPod *corev1.Pod
+
 	flag.Parse()
+
 	kubeClient := client.NewKubeClientSet()
+	nrClient := client.NewNodeRecoveryClientSet()
+	nodeConditionManager := controller.NewNodeConditionManager()
+
+	BeforeEach(func() {
+		By("Getting master node")
+		masterNode, err := tests.GetMasterNode()
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Creating fake IPMI server pod")
+		_, err = tests.CreateFakeIpmiPod(masterNode.Name)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Creating service for fake IPMI server")
+		_, err = tests.CreateFakeIpmiService(tests.ServiceFakeIpmiClusterIP, tests.ServiceFakeIpmiPort)
+		Expect(err).ToNot(HaveOccurred())
 
-	It("test", func() {
-		_, err := kubeClient.CoreV1().Pods("default").List(metav1.ListOptions{})
+		By("Creating pod to execute SSH commands")
+		sshExecPod, err = tests.CreateSSHExecPod(masterNode.Name)
 		Expect(err).ToNot(HaveOccurred())
+	})
+
+	When("node has \"NotReady\" state", func() {
+		BeforeEach(func() {
+			By("Stoping kubelet service on the non-master node")
+			Eventually(
+				func() error {
+					_, _, err := tests.RunSSHCommand(
+						sshExecPod,
+						tests.NonMasterNode,
+						tests.NodeUser,
+						[]string{"sudo", "systemctl", "stop", "kubelet"},
+					)
+					return err
+				}, 60*time.Second, time.Second,
+			).ShouldNot(HaveOccurred())
+
+			By("Waiting until non-master node will have \"NonReady\" state")
+			Eventually(
+				func() bool {
+					node, err := kubeClient.CoreV1().Nodes().Get(tests.NonMasterNode, metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+
+					readyCond := nodeConditionManager.GetNodeCondition(node, corev1.NodeReady)
+					return readyCond.Status == corev1.ConditionUnknown
+				}, 90*time.Second, 5*time.Second,
+			).Should(BeTrue())
+		})
+
+		It("should remediate the node", func() {
+			By("Checking that node remediation object was created")
+			Eventually(
+				func() *v1.NodeRemediation {
+					nr, err := nrClient.NoderecoveryV1alpha1().NodeRemediations().Get(tests.NonMasterNode, metav1.GetOptions{})
+					if err != nil {
+						return nil
+					}
+					return nr
+				}, 30*time.Second, 1*time.Second,
+			).ShouldNot(BeNil())
+		})
+
+		AfterEach(func() {
+			By("Starting kubelet service on the non-master node")
```

Comment:
Stupid question... why do you need to start it again?

Reply:
@beekhof It does not really restart the node via IPMI, so `kubelet` service will stay down forever.
To be sure that I will have the same environment for each test I start it again.
I also wanted to check `provider-external` jobs, to see if it succeeds, but I will do it in the separate PR, after we will finish discussion around https://github.com/kubevirt/cluster-api-provider-external/pull/6.